Handle zero contour aligning with faces in MeshClassifier.

simonsticko opened the pull request:
Change logic in `MeshClassifier` to handle the situation described  issue #16888. Add a new type to the enum `LocationToLevelSet` that  describes that the zero contour is exactly aligned with a face. Change the logic in `MeshClassifier` to set the "inside" cell sharing an aligned face to intersected and the other cell to outside.

simonsticko commented:
I was thinking something like this @mschreter. Do you think this makes sense?

mschreter commented on include/deal.II/non_matching/mesh_classifier.h:
```diff
@@ -56,7 +57,8 @@ namespace NonMatching
     intersected,
-    unassigned
+    unassigned,
+    aligned
```
Shall we add a comment/assert that a cell should never be aligned?

simonsticko replied:
I think there should be an assert. I will add it.

mschreter commented on source/non_matching/mesh_classifier.cc:
```diff
@@ -68,9 +68,11 @@ namespace NonMatching
                               local_levelset_values.end());
 
-        if (*min_max_element.second < 0)
+        if (*min_max_element.first == 0.0 && *min_max_element.second == 0.0)
```
Should we introduce a small tolerance for the check?

simonsticko replied:
I agree that, usually, one should never compare floating point number exactly. But here, also I think that we actually want to. Moving the zero-contour would just change it to one of the other values of `LocationToLevelSet`.

Regarding tracefem: If we choose to implement it like this I do not think you need to perturb the level set. If a face is aligned, you integrate over the zero contour of the cell that is chosen as intersected and not over the other cell, so it will work also for this case.

mschreter reviewed (commented):
@simonsticko looks very good. I've tried the patch for our application considered in #16888 and the `active_fe_index` now looks as we agreed. Thank you very much :-).

mschreter commented:
@simonsticko even though some time has passed, do you think it’s worth revisiting and trying to get this feature merged? I would be happy to help address any remaining points if needed. We currently have an open PR in our user code with a workaround that I would like to resolve one way or another. Thank you :-).

simonsticko commented:
Thank you @mschreter! Sorry for not fixing this sooner...

I have now

- I applied your patch, and added some more documentation that I had stared writing at some point.
  I added you as co-author on the commit.
- Made two minor logic changes:
     - Check first if there are any intersected faces, before checking for aligned faces, since this is the more common scenario.
     - Added two asserts
- Added a change-log entry.

I would be happy to get this merged now, as soon as anyone has reviewed it.

masterleinad commented:
/rebuild
